src/main/java/games/chess/components/ChessPiece.java

```diff
@@ -0,0 +1,137 @@
+package games.chess.components;
+
+import java.util.Objects;
+
+import core.components.Token;
+
+
+public class ChessPiece extends Token {
+
+    public enum ChessPieceType {
+        KING,
+        QUEEN,
+        ROOK,
+        BISHOP,
+        KNIGHT,
+        PAWN
+    }
+
+    //Type of the piece (e.g. KING, QUEEN, etc.)
+    private ChessPieceType type;
+
+
+    public enum MovedState {
+        NOT_MOVED,
+        MOVED, 
+        NOT_RELEVANT
+    }
+    
+    //Wether the piece has moved or not (e.g. for castling, en passant, etc.), only relevant for pawns, kings and rooks
+    private MovedState movedState;
+
+    private int x;
+    private int y;
```

Comment:
You are storing the position of a piece on the ChessPiece, and also recording the ChessPiece on an (x, y) position on the ChessBoard.

Both of these are individually very reasonable, but you have the same information stored in two separate places, which opens up all sorts of possible subtle bugs if the two sets of information get out of synch with each other.

This may be a refactor too far, but consider if there is a way to have a single source of the information. For example, could you make ChessPiece immutable and just have the mutable information stored on the chessboard.

Reply:
Perfectly reasonable conclusion.